core/src/main/scala/algebra/lattice/GenBool.scala

```diff
@@ -0,0 +1,57 @@
+package algebra
+package lattice
+
+import ring.BoolRng
+import scala.{specialized => sp}
+
+/**
+ * Generalized Boolean algebra, that is, a Boolean algebra without
+ * the top element. Generalized Boolean algebras do not (in general)
+ * have (absolute) complements, but they have ''relative complements''
+ * (see [[GenBool.without]]).
+ */
+trait GenBool[@sp(Int, Long) A] extends Any with DistributiveLattice[A] with BoundedJoinSemilattice[A] { self =>
+  def and(a: A, b: A): A
+  override def meet(a: A, b: A): A = and(a, b)
+
+  def or(a: A, b: A): A
+  override def join(a: A, b: A): A = or(a, b)
+
+  /**
+   * The operation of ''relative complement'', symbolically often denoted
+   * `a\b` (the symbol for set-theoretic difference, which is the
+   * meaning of relative complement in the lattice of sets).
+   */
+  def without(a: A, b: A): A
+
+  /**
+   * Logical exclusive or, set-theoretic symmetric difference.
+   * Defined as `a\b ∨ b\a`.
+   */
+  def xor(a: A, b: A): A = or(without(a, b), without(b, a))
+
+  /**
+   * Every generalized Boolean algebra is also a `BoolRng`, with
+   * multiplication defined as `and` and addition defined as `xor`.
+   */
+  def asBoolRing: BoolRng[A] = new BoolRngFromGenBool(self)
```

Comment:
This was intentional, because I didn't want to introduce a new method name.
It is kind of analogous to dual, where the subtype overrides the return
type.
On Oct 30, 2015 9:25 PM, "P. Oscar Boykin" <notifications@github.com> wrote:

> In core/src/main/scala/algebra/lattice/GenBool.scala
> <https://github.com/non/algebra/pull/109#discussion_r43566192>:
>
> > +   * `a\b` (the symbol for set-theoretic difference, which is the
> > +   * meaning of relative complement in the lattice of sets).
> > +   */
> > +  def without(a: A, b: A): A
> > +
> > +  /**
> > +   * Logical exclusive or, set-theoretic symmetric difference.
> > +   * Defined as `a\b ∨ b\a`.
> > +   */
> > +  def xor(a: A, b: A): A = or(without(a, b), without(b, a))
> > +
> > +  /**
> > +   * Every generalized Boolean algebra is also a `BoolRng`, with
> > +   * multiplication defined as `and` and addition defined as `xor`.
> > +   */
> > +  def asBoolRing: BoolRng[A] = new BoolRngFromGenBool(self)
>
> asBoolRng, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/non/algebra/pull/109/files#r43566192>.
>